src/reflect/scala/reflect/runtime/JavaMirrors.scala

```diff
@@ -423,12 +421,20 @@ private[scala] trait JavaMirrors extends internal.SymbolTable with api.JavaUnive
     // caches MethodSymbol metadata, so that we minimize the work that needs to be done during Mirror.apply
     // TODO: vararg is only supported in the last parameter list (scala/bug#6182), so we don't need to worry about the rest for now
     private class MethodMetadata(symbol: MethodSymbol) {
-      private val params = symbol.paramss.flatten.toArray
-      private val vcMetadata = params.map(p => new DerivedValueClassMetadata(p.info))
-      val isByName = params.map(p => isByNameParam(p.info))
-      def isDerivedValueClass(i: Int) = vcMetadata(i).isDerivedValueClass
+      val paramCount = sumSize(symbol.paramss, 0)
+      private val vcMetadata = new Array[DerivedValueClassMetadata](paramCount)
+      val isByName: BitSet = new BitSet(initSize = paramCount)
+      {
+        var ix = 0 ; val iter: Iterator[Symbol] = symbol.paramss.iterator.flatten
+        while (iter.hasNext && ix < paramCount){
```

Comment:
`ix < paramCount` looks like it would always be true, can we just rely on exhausting the iterator?

Reply:
Yes, `0 <= ix < paramCount` is an invariant of the loop, which ensures that array accesses are in-bounds. Sometimes I find it convenient to explicitly put the invariant in the condition.

Reply:
I found it hurts readability somewhat: I need to ask myself if there is more than one way to stop the loop.

`.zipWithIndex.foreach` would be easier to read, IMO, albeit at the cost of some temporary tuples. We could add an custom combinator `foreachWithIndex[T](Iterator[T])((T, Int) => Unit)` to avoid that cost.

Reply:
Ok, I think it can be removed then.

Reply:
I ended up adding that `foreachWithIndex` method to the `Collections` trait, and using it here.